core/src/main/java/com/linecorp/armeria/client/Clients.java

```diff
@@ -581,18 +582,51 @@ public static SafeCloseable withHttpHeaders(Function<HttpHeaders, HttpHeaders> h
      *
      * @see #withHttpHeaders(Function)
      */
-    public static SafeCloseable withContextCustomizer(Consumer<ClientRequestContext> contextCustomizer) {
+    public static SafeCloseable withContextCustomizer(
+            Consumer<? super ClientRequestContext> contextCustomizer) {
         requireNonNull(contextCustomizer, "contextCustomizer");
 
-        final Consumer<ClientRequestContext> oldCustomizer = THREAD_LOCAL_CONTEXT_CUSTOMIZER.get();
+        final ClientThreadLocalState customizers = maybeCreateContextCustomizers();
+        customizers.add(contextCustomizer);
 
-        if (oldCustomizer != null) {
-            THREAD_LOCAL_CONTEXT_CUSTOMIZER.set(oldCustomizer.andThen(contextCustomizer));
-            return () -> THREAD_LOCAL_CONTEXT_CUSTOMIZER.set(oldCustomizer);
-        } else {
-            THREAD_LOCAL_CONTEXT_CUSTOMIZER.set(contextCustomizer);
-            return THREAD_LOCAL_CONTEXT_CUSTOMIZER::remove;
+        return new SafeCloseable() {
+            boolean closed;
+
+            @Override
+            public void close() {
+                if (closed) {
+                    return;
+                }
+
+                closed = true;
+                customizers.remove(contextCustomizer);
```

Comment:
Should we clean up threadlocal if no more customizers?

Reply:
Good point. Yeah, we should because otherwise we may leak in a container.

Reply:
Actually, that wouldn't prevent a leak. A user has to make sure to close all `SafeCloseable`s and make a call to clear a pending captor.

Reply:
That being said, I think it's safe to leave the thread-local state.

Reply:
Just realized that the thread will still have a reference to `ClientThreadLocalState` instance itself, so it's not really safe. :sweat_smile:

Reply:
Done

Reply:
It's normal in Java to require all closeables to be closed to prevent leaks - Java doesn't have proper RAII unfortunately and try/resources is just ok but is all we've got.

If we have a case here where `Captor.get` needs to be called to prevent a leak, I really feel instead it should be `SafeCloseable` instead of `Supplier` to make this explicit and allow users to use try/resources which is safe even with exceptions. It's a little more verbose sure, but it's just how Java is (Java 10+ help a lot with `var` and have far less of the burden anyways). 

Anyways, if you feel strongly about the current API I'm fine with merging but want to point this out.

Reply:
> If we have a case here where `Captor.get` needs to be called to prevent a leak, I really feel instead it should be `SafeCloseable` instead of `Supplier` to make this explicit and allow users to use try/resources which is safe even with exceptions. 

Agreed. A user currently must call `Captor.get()` to prevent a leak in the following cases:

- A user specified an illegal request path.
- A user specified a request URI without an authority when a `WebClient` was created without the default authority.

We could 1) make it `SafeCloseable` or 2) make sure a context is created for all cases and consider it's a bug if a context is not created.

Reply:
+1 for `SafeCloseable`. If I'm not mistaken, there are more cases for exceptions since it seems reasonable for a user to call aggregate.join on the same line, or they could be using a blocking stub.

Reply:
Thread local will be cleaned up as soon as a new context is ceeated, so aggregate() or anything that happens after context creation is irrelevant.

Reply:
I think it's better making it `SafeCloseable` because we can't have complete control over what exceptions are gonna be triggered before a context is created. Please check the new API: https://github.com/line/armeria/pull/2344/commits/a3e4be169cde43af4adb10f8144a93be4d3992c2